### ISSUE TITLE
Multi-threading issue

### ISSUE THREAD

**Issue by jbaron** [PhilipVinc/TensorBoardLogger.jl#72]
I'm trying to use TensorBoardLogger in a multi-threaded environment. Every thread gets its own logger, writing to a different file. But I'm still getting occasional exceptions like the one below (ProtoBuf related ???).

Is there anything else that I should keep in mind?

```
TaskFailedException:
BoundsError: attempt to access 2×8 BitArray{2} at index [1:2, 9]
Stacktrace:
 [1] throw_boundserror(::BitArray{2}, ::Tuple{UnitRange{Int64},Int64}) at ./abstractarray.jl:537
 [2] checkbounds at ./abstractarray.jl:502 [inlined]
 [3] view at ./subarray.jl:155 [inlined]
 [4] maybeview at ./views.jl:133 [inlined]
 [5] dotview at ./broadcast.jl:1160 [inlined]
 [6] _unset_oneofs(::BitArray{2}, ::Array{Int64,1}, ::Int64) at /home/peter/.julia/packages/ProtoBuf/JY3fQ/src/codec.jl:612
 [7] _fillset(::TensorBoardLogger.Summary_Value, ::Symbol, ::Bool, ::Bool) at /home/peter/.julia/packages/ProtoBuf/JY3fQ/src/codec.jl:641
 [8] fillset at /home/peter/.julia/packages/ProtoBuf/JY3fQ/src/codec.jl:633 [inlined]
 [9] setproperty! at /home/peter/.julia/packages/ProtoBuf/JY3fQ/src/utils.jl:5 [inlined]
 [10] _protobuild(::TensorBoardLogger.Summary_Value, ::Base.Iterators.Pairs{Symbol,Any,Tuple{Symbol,Symbol},NamedTuple{(:tag, :simple_value),Tuple{String,Float32}}}) at /home/peter/.julia/packages/ProtoBuf/JY3fQ/src/utils.jl:61
 [11] Summary_Value at /home/peter/.julia/packages/TensorBoardLogger/gv4oF/src/protojl/summary_pb.jl:63 [inlined]
 [12] scalar_summary at /home/peter/.julia/packages/TensorBoardLogger/gv4oF/src/Loggers/LogValue.jl:19 [inlined]
 [13] log_metric(::TensorBoardLogger.TBLogger, ::Symbol, ::Float32, ::Int64, ::DateTime) at /home/peter/dev/Quanta.jl/test/multi.jl:5
 [14] macro expansion at /home/peter/dev/Quanta.jl/test/multi.jl:21 [inlined]
 [15] (::var"#29#threadsfor_fun#8"{UnitRange{Int64}})(::Bool) at ./threadingconstructs.jl:81
 [16] (::var"#29#threadsfor_fun#8"{UnitRange{Int64}})() at ./threadingconstructs.jl:48
wait at task.jl:267 [inlined]
threading_run(::Function) at threadingconstructs.jl:34
macro expansion at threadingconstructs.jl:93 [inlined]
test123() at multi.jl:15
top-level scope at multi.jl:27
include_string(::Function, ::Module, ::String, ::String) at loading.jl:1088
```

**Comment by jbaron** [comment]
This code can reproduce the issue for anyone who is interested:

```
import TensorBoardLogger
using Dates

function log_metric(logger::TensorBoardLogger.TBLogger, name::Symbol, value::Real, step::Int, time::DateTime)
    summary = TensorBoardLogger.SummaryCollection(
        TensorBoardLogger.scalar_summary(string(name), value)
    )
    event = TensorBoardLogger.Event(wall_time=Dates.datetime2unix(time), summary=summary, step=step)
    TensorBoardLogger.write_event(logger.file, event)
end


function test123()
    Threads.@threads for i in 1:500
        logger = TensorBoardLogger.TBLogger("tmp/run_" * string(i))
        now = Dates.now()
        for step in 1:10_000
            log_metric(logger, :test, 12.0f0, step, now)
            sleep(0.001)
        end
    end
end

test123()

```

**Comment by oxinabox** [comment]
Looks protobuf related to me.
cc @tanmaykm

**Comment by jbaron** [comment]
Based on the stacktrace indeed looks like that. And if it is a bug (and not expected behaviour, like tensorboardlogger cannot be used in a multithreaded env), feel free to close it and I can always re-open an issue against protobuf.

**Comment by tanmaykm** [comment]
Yes. ProtoBuf.jl keeps some state in a common dict. It has not been made thread safe yet.

**Comment by jbaron** [comment]
Good to know. I guess for the time being I can always serialize access to the logger. Shall I close this issue?

**Comment by PhilipVinc** [comment]
HI @jbaron and thanks for the Bug Report.
Indeed I had never tested Tensorboard for multithreading. There would most probably also be issues when writing to file.

I'd leave this issue open for reference. 

A more minimal MWE is
```julia
using TensorBoardLogger
function test123()
           Threads.@threads for i in 1:500
               for step in 1:10_000
                   TensorBoardLogger.Summary_Value(tag="ciao", simple_value=3.2)
                   sleep(0.001)
               end
           end
       end

test123()
```

@tanmaykm the error arises on `_protobuild`, but I fail to see where there is shared state in that codepath... 
what common dicts are you talking about?

**Comment by tanmaykm** [comment]
I was referring to [these](https://github.com/JuliaIO/ProtoBuf.jl/blob/06a932a795a1676c39513a29b1be56a465cf8514/src/codec.jl#L534-L536).

**Comment by PhilipVinc** [comment]
@tanmaykm from what I understand those are used to track what fields of a protobuf structure are set to a non-default non-null value? Probably we should move this discussion over to ProtoBuf.jl, but did you ever consider alternative strategies, such as moving those structures to a field of the struct itself, or something like this?

**Comment by tanmaykm** [comment]
Yes indeed. I think moving that into the struct and having `getproperty` - `setproperty` would solve this. I have opened this issue (https://github.com/JuliaIO/ProtoBuf.jl/issues/140) on ProtoBuf.jl for this.

**Comment by PhilipVinc** [comment]
Closed by #80
